### DISCUSSION TITLE
To support multiple ports for autoscaling

### DISCUSSION THREAD

**Discussion by palashbiswas-git** [apache/cloudstack#9152] [Ideas]
##### ISSUE TYPE

 * Bug Report
 * Improvement Request
 * Enhancement Request

##### COMPONENT NAME

~~~
Autoscaling
~~~

##### CLOUDSTACK VERSION

~~~
4.19.0.0
~~~

##### SUMMARY

~~~
Could you please enhance CloudStack to enable support for multiple ports in autoscaling? The current limitation of only one port restricts the deployment of real-time production infrastructure with autoscaling groups.
~~~

##### EXPECTED RESULTS

~~~
To address this issue of limited ports in the next release!
~~~

##### ACTUAL RESULTS

~~~

~~~

**Comment by DaanHoogland** [comment]
at first sight this looks like a valid feature request @palashbiswas-git , but can you elaborate on the configurations/requests/settings that you expect and how they would bahave, including error handling.

**Comment by palashbiswas-git** [comment]
Hi @DaanHoogland ,

I'm currently working with a container-based application that supports autoscaling. However, I'm facing a limitation due to having a single load balancer and single port, which prevents me from parking it under an Autoscaling group, especially when I have multiple containers in a single VM/Instance. As you know, except for frontend web-based applications, most applications work with multiple ports where the VM needs to connect to other corresponding nodes, containers, and other services.

Because of this limitation, I believe many Cloudstack users are unable to secure those backend critical systems/applications through autoscaling groups.

Therefore, it would be greatly appreciated if this feature could be included in an upcoming enhancement.

I hope this clarifies the issue and the configuration/setting we are looking for in the future.

Thank you.

**Comment by kiranchavala** [comment]
Hi @palashbiswas-git 

Have you tried cloudstack autoscaler for kubernetes ?


https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler/cloudprovider/cloudstack


When a Kubernetes pod scales up , cloudstack automatically provisions a node 



https://github.com/apache/cloudstack-kubernetes-provider/issues/52#issuecomment-1968230660

**Comment by palashbiswas-git** [comment]
Hi @kiranchavala,

You are correct that this can be achieved through Kubernetes. However, our requirement is to use either containers or normal VMs, where we can't handle multiple load balancers or a single load balancer with multiple ports if it's part of an autoscaling group.

**Comment by btzq** [comment]
This is a great feature for us. Right now, in order to overcome this limitation, we deploy an NGINX instance in our VM, configured with static routes. These static routes would then route traffic to the multiple dockers in the VM. 

Then, we scale the entire VM altogether.

Its a pretty hacky workaround, but it helps simplify our deployment. 

Autoscaling with multiple ports would be really great! I full support.

**Comment by DaanHoogland** [comment]
@palashbiswas-git , I am sure you have a valid use case, and I think something like this could be implemented in the future. I read this as a generic problem statement and not a very explicit one. I am moving this to discussion to further distill it and create (probably multiple) issues once it is more clear.

**Comment by btzq** [comment]
Hey @DaanHoogland , this feature would be really useful for us. 

We Autoscale a bunch of API Gateways, and have to maintain 1 ASG for Each Port. 

If there is a way to support multiple ports in the same ASG Group, then it would save a lot of resources and duplicate VMs on our side (and public IPs). 

I see that there are related discussions about this:
- https://github.com/apache/cloudstack/issues/8313

Will #8313 resolve this?
